nmigen/build/run.py

```diff
@@ -99,6 +100,64 @@ def execute_local(self, root="build", *, run_script=True):
         finally:
             os.chdir(cwd)
 
+    def execute_remote_ssh(self, root, hostname, *, connect_args = {}):
+        """
+        Execute build plan using the remote SSH strategy. Files from the build
+        plan are transferred via SFTP to the directory ``root`` on the remote
+        server ``hostname``. The ``paramiko`` SSH client will then run ``{script}.sh``.
+
+        ``hostname`` corresponds to the first (required) input argument of ``paramiko``'s
+        ``SSHClient.connect``, and ``connect_args`` is a dictionary that holds
+        all other input arguments to ``SSHClient.connect``
+        (`documentation <http://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.connect>`_).
+
+        Returns :class:`RemoteSshBuildProducts`.
+        """
+        from paramiko import SSHClient
+
+        with SSHClient() as client:
+            client.load_system_host_keys()
+            client.connect(hostname, **connect_args)
+
+            with client.open_sftp() as sftp:
+                try:
+                    sftp.mkdir(root)
+                except IOError as e:
+                    pass # mkdir fails if directory exists. This is fine.
+
+                sftp.chdir(root)
+                for filename, content in self.files.items():
+                    filename = os.path.normpath(filename)
+                    # Just to make sure we don't accidentally overwrite anything outside of build root.
+                    assert not filename.startswith("..")
+
+                    dirname = os.path.dirname(filename)
+                    if dirname:
+                        try:
+                            sftp.mkdir(dirname, exist_ok=True)
+                        except IOError as e:
+                            pass
+
+                    mode = "wt" if isinstance(content, str) else "wb"
+                    with sftp.file(filename, mode) as f:
+                        f.write(content)
+
+            cmd = "cd {} && bash -l {}.sh".format(root, self.script)
+            stdin, stdout, stderr = client.exec_command(cmd)
+
+            # Show the output from the server while products are built.
```

Comment:
This seems like it could lead to deadlocks if something first writes a lot to stderr but not to stdout, but I'm not sure.

Reply:
`stdout` and `stderr` are subclassed from `paramiko`'s `BufferedFile`- it's called `ChannelFile`.

I can portably (Windows and *nix) `select` on them by accessing their [`channel`](https://github.com/paramiko/paramiko/blob/master/paramiko/channel.py#L1350). Based on the docstring that says `ChannelFile` wraps a `Channel`, I think accessing the `channel` field directly should be supported?

Reply:
Ah. I think in this case it would be more appropriate to do something like:
```
transport = client.get_transport()
channel = transport.open_session()
channel.set_combine_stderr(True)
channel.exec_command("...")
```

Then you can poll just stdout.